File: apps/backend/classifications/questionnaire/scoring/engine.py
"""
Scoring accumulator & ratio normalizer — SBGC-174 (Epic SBGC-171).

Pure functions:

* :func:`compute_raw_profile` — accumulates per-step-floored dimension scores
  from answered option modifiers (a negative modifier never creates a negative
  running balance; later positives build up from 0).
* :func:`normalize_profile` — ratio-normalizes a raw profile to a 100-point
  integer distribution using the Largest Remainder method with a fixed
  tie-break order (Micro ≻ Macro ≻ Mystiko) and a zero-total fallback of
  ``(33, 33, 34)``.
"""

from __future__ import annotations

import math
from collections.abc import Sequence

from classifications.questionnaire.registry.v1.types import (
    AnswerOption,
    ProfileTarget,
    QuestionNode,
)
from classifications.questionnaire.scoring.types import DimensionScore

#: Static dimension tie-break priority (Micro ≻ Macro ≻ Mystiko).
_DIMENSION_PRIORITY = ("micro", "macro", "mystiko")


def compute_raw_profile(
    answers: dict[str, str],
    nodes: Sequence[QuestionNode],
    target: ProfileTarget,
) -> DimensionScore:
    """Accumulate *answers* into a raw profile with per-step zero-flooring.

    *answers* maps a question node id to its selected option id.  Questions
    whose ``target`` differs from *target* are ignored, and unknown
    question/option ids are skipped (the submission validator owns hard
    rejection — SBGC-176).
    """
    micro = 0
    macro = 0
    mystiko = 0

    node_map = {node.id: node for node in nodes if node.target is target}

    for question_id, option_id in answers.items():
        node = node_map.get(question_id)
        if node is None:
            continue
        option: AnswerOption | None = next(
            (candidate for candidate in node.options if candidate.id == option_id),
            None,
        )
        if option is None:
            continue

        micro = max(0, micro + option.modifiers.micro)
        macro = max(0, macro + option.modifiers.macro)
        mystiko = max(0, mystiko + option.modifiers.mystiko)

    return DimensionScore(micro=micro, macro=macro, mystiko=mystiko)


def normalize_profile(raw: DimensionScore) -> DimensionScore:
    """Normalize *raw* to a 100-point integer distribution (Hamilton–Hare)."""
    total = raw.total
    if total == 0:
        return DimensionScore(micro=33, macro=33, mystiko=34)

    raw_values = {
        "micro": raw.micro,
        "macro": raw.macro,
        "mystiko": raw.mystiko,
    }
    exact = {
        dimension: (value / total) * 100.0 for dimension, value in raw_values.items()
    }
    floors = {
        dimension: math.floor(exact[dimension]) for dimension in _DIMENSION_PRIORITY
    }

    remainder = 100 - sum(floors.values())

    # Sort by: fractional remainder (desc), then raw magnitude (desc), then the
    # static dimension priority.  ``priority`` is a negative index so a higher
    # priority dimension sorts earlier under ``reverse=True``.
    fractions = [
        (
            exact[dimension] - floors[dimension],
            raw_values[dimension],
            -index,
            dimension,
        )
        for index, dimension in enumerate(_DIMENSION_PRIORITY)
    ]
    fractions.sort(key=lambda item: (item[0], item[1], item[2]), reverse=True)

    allocations = dict(floors)
    for index in range(remainder):
        allocations[fractions[index][3]] += 1

    return DimensionScore(
        micro=allocations["micro"],
        macro=allocations["macro"],
        mystiko=allocations["mystiko"],
    )


__all__ = ["compute_raw_profile", "normalize_profile"]

File: apps/frontend/src/lib/questionnaire/scoring/engine.ts
/**
 * Scoring accumulator & ratio normalizer — SBGC-174 (Epic SBGC-171).
 *
 * Mirrors `apps/backend/classifications/questionnaire/scoring/engine.py`.
 */

import type {
  AnswerOption,
  ProfileTarget,
  QuestionNode,
} from "../registry/v1/types";
import type { DimensionScore } from "./types";

const DIMENSION_PRIORITY: readonly (keyof DimensionScore)[] = [
  "micro",
  "macro",
  "mystiko",
];

/** Accumulate answers into a raw profile with per-step zero-flooring. */
export function computeRawProfile(
  answers: Record<string, string>,
  nodes: readonly QuestionNode[],
  target: ProfileTarget,
): DimensionScore {
  let micro = 0;
  let macro = 0;
  let mystiko = 0;

  const nodeMap = new Map(
    nodes
      .filter((node) => node.target === target)
      .map((node) => [node.id, node]),
  );

  for (const [questionId, optionId] of Object.entries(answers)) {
    const node = nodeMap.get(questionId);
    if (node === undefined) continue;
    const option: AnswerOption | undefined = node.options.find(
      (candidate) => candidate.id === optionId,
    );
    if (option === undefined) continue;

    micro = Math.max(0, micro + option.modifiers.micro);
    macro = Math.max(0, macro + option.modifiers.macro);
    mystiko = Math.max(0, mystiko + option.modifiers.mystiko);
  }

  return { micro, macro, mystiko };
}

/** Normalize a raw profile to a 100-point integer distribution (Largest Remainder). */
export function normalizeProfile(raw: DimensionScore): DimensionScore {
  const total = raw.micro + raw.macro + raw.mystiko;
  if (total === 0) {
    return { micro: 33, macro: 33, mystiko: 34 };
  }

  const values: Record<keyof DimensionScore, number> = {
    micro: raw.micro,
    macro: raw.macro,
    mystiko: raw.mystiko,
  };
  const exact: Record<keyof DimensionScore, number> = {
    micro: (values.micro / total) * 100,
    macro: (values.macro / total) * 100,
    mystiko: (values.mystiko / total) * 100,
  };
  const floors: Record<keyof DimensionScore, number> = {
    micro: Math.floor(exact.micro),
    macro: Math.floor(exact.macro),
    mystiko: Math.floor(exact.mystiko),
  };

  const remainder = 100 - (floors.micro + floors.macro + floors.mystiko);

  const fractions: Array<{
    fraction: number;
    raw: number;
    priority: number;
    dimension: keyof DimensionScore;
  }> = DIMENSION_PRIORITY.map((dimension, index) => ({
    fraction: exact[dimension] - floors[dimension],
    raw: values[dimension],
    priority: -index,
    dimension,
  }));

  // Descending by fraction, then raw magnitude, then static priority.
  fractions.sort(
    (a, b) =>
      b.fraction - a.fraction || b.raw - a.raw || b.priority - a.priority,
  );

  const allocations: Record<keyof DimensionScore, number> = { ...floors };
  for (let index = 0; index < remainder; index += 1) {
    allocations[fractions[index].dimension] += 1;
  }

  return {
    micro: allocations.micro,
    macro: allocations.macro,
    mystiko: allocations.mystiko,
  };
}
